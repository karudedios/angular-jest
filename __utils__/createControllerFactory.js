export default function(controllerName, cb) {

  return function($componentController, $rootScope, $compile) {

    cb(function(params) {

      const controller = $componentController(controllerName, null, params);

      controller.$onInit && controller.$onInit();

      return controller;

    }, function(elementName, props = {}) {
        const scope = $rootScope.$new();
        const keys = Object.keys(props);
        Object.assign(scope, props);

        const htmlElementName = camelCaseToDashCase(elementName);

        const htmlProps = keys.map(key => {
          return `${camelCaseToDashCase(key)}="${key}"`
        }, '').join(' ');

        const element = $compile(`<${htmlElementName} ${htmlProps}></${htmlElementName}>`)(scope);

        scope.$apply(); // Make sure we actually perform any HTML change

        return element;

        function camelCaseToDashCase(str) {
          return str.replace(/[A-Z]/g, $1 => `-${$1.toLowerCase()}`);
        }

    });

  };

}
