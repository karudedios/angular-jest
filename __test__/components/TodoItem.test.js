import * as R from 'ramda';
import createControllerFactory from '../../__utils__/createControllerFactory';

describe('TodoItem', function() {

  let createController;

  beforeEach(() => {

    angular.mock.module('jest-webpack-app.components');

  });

  beforeEach(inject(createControllerFactory('todoItem', creator => {

    createController = creator;

  })));

  it('should default to an empty object when no todo is provided', function() {

    const controller = createController();

    expect(controller.todo).toBeDefined();
    expect(controller.todo).toMatchObject({});

  });

  it('should call `onChangeRequest` function with proper payload when `update` is called', function() {

    const todo = { title: "todo title" };

    const controller = createController({
      todo,
      onChangeRequest: jest.fn(),
    });

    const updatedTodo = R.set(R.lensProp('completed'), true, controller.todo);

    controller.toggleTodo();

    expect(controller.onChangeRequest).toBeCalledWith({ $todo: updatedTodo });

  });

  it('should call `onRemoveRequest` function with proper payload when `delete` is called', function() {

    const todo = { title: 'todo title' };

    const controller = createController({
      todo,
      onRemoveRequest: jest.fn()
    });

    controller.removeTodo();

    expect(controller.onRemoveRequest).toBeCalledWith({ $todo: controller.todo });

  });

});
