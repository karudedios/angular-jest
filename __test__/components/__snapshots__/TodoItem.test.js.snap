// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Snapshot Testing should render the element when no todo is provided 1`] = `
<todo-item
  class="ng-scope ng-isolate-scope"
>
  
    
  <div
    class="todo"
  >
    
      
    <input
      data-ng-checked="todoItem.todo.completed"
      id="check-"
      type="checkbox"
    />
    

      
    <label
      class="todo-title"
      data-ng-click="::todoItem.toggleTodo()"
    >
      
        
      <h4
        class="ng-binding"
        data-ng-bind="::todoItem.todo.title"
      />
      
      
    </label>
    

      
    <span
      class="svg-icon delete-icon action small"
      data-ng-click="::todoItem.removeTodo()"
    >
      
        svg-icon
      
    </span>
    
    
  </div>
  
  
</todo-item>
`;

exports[`TodoItem Snapshot Testing should render the element when todo is provided with a title 1`] = `
<todo-item
  class="ng-scope ng-isolate-scope"
  todo="todo"
>
  
    
  <div
    class="todo"
  >
    
      
    <input
      data-ng-checked="todoItem.todo.completed"
      id="check-"
      type="checkbox"
    />
    

      
    <label
      class="todo-title"
      data-ng-click="::todoItem.toggleTodo()"
    >
      
        
      <h4
        class="ng-binding"
        data-ng-bind="::todoItem.todo.title"
      >
        Hello!
      </h4>
      
      
    </label>
    

      
    <span
      class="svg-icon delete-icon action small"
      data-ng-click="::todoItem.removeTodo()"
    >
      
        svg-icon
      
    </span>
    
    
  </div>
  
  
</todo-item>
`;

exports[`TodoItem Snapshot Testing should render the element when todo is provided with id 1`] = `
<todo-item
  class="ng-scope ng-isolate-scope"
  todo="todo"
>
  
    
  <div
    class="todo"
  >
    
      
    <input
      data-ng-checked="todoItem.todo.completed"
      id="check-5"
      type="checkbox"
    />
    

      
    <label
      class="todo-title"
      data-ng-click="::todoItem.toggleTodo()"
    >
      
        
      <h4
        class="ng-binding"
        data-ng-bind="::todoItem.todo.title"
      />
      
      
    </label>
    

      
    <span
      class="svg-icon delete-icon action small"
      data-ng-click="::todoItem.removeTodo()"
    >
      
        svg-icon
      
    </span>
    
    
  </div>
  
  
</todo-item>
`;
